src/librustc_typeck/check/demand.rs

```diff
@@ -374,6 +379,77 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
         None
     }
 
+    // This function checks if the specified expression is a built-in range literal
+    // (See: librustc/hir/lowering.rs::LoweringContext::lower_expr() )
+    fn is_range_literal(&self, expr: &hir::Expr) -> bool {
+        use hir::{Path, QPath, ExprKind, TyKind};
+
+        // we support `::std::ops::Range` and `::std::core::Range` prefixes
+        // (via split on "|")
+        let ops_path = ["{{root}}", "std|core", "ops"];
+
+        let is_range_path = |path: &Path| {
+            let ident_names: Vec<_> = path.segments
+                .iter()
+                .map(|seg| seg.ident.as_str())
+                .collect();
+
+            if let Some((last, preceding)) = ident_names.split_last() {
+                last.starts_with("Range") &&
+                    preceding.len() == 3 &&
+                    preceding.iter()
+                        .zip(ops_path.iter())
+                        .all(|(seg, match_seg)| {
+                            match_seg.split("|")
+                                .into_iter()
+                                .any(|ref spl_seg| seg == spl_seg)
+                        })
+            } else {
+                false
+            }
+        };
+
+        let is_range_struct_snippet = |span: &Span| {
+            // Tell if expression span snippet looks like an explicit
+            // Range struct or new() call.  This is to allow rejecting
+            // Ranges constructed with non-literals.
+            let source_map = self.tcx.sess.source_map();
+            let end_point = source_map.end_point(*span);
+
+            if let Ok(end_string) = source_map.span_to_snippet(end_point) {
+                end_string.ends_with("}") || end_string.ends_with(")")
+            } else {
+                false
+            }
+
+        };
+
+        match expr.node {
+            // all built-in range literals but `..=` and `..`
+            // desugar to Structs, `..` desugars to its struct path
+            ExprKind::Struct(QPath::Resolved(None, ref path), _, _) |
+            ExprKind::Path(QPath::Resolved(None, ref path)) => {
+                return is_range_path(&path) && !is_range_struct_snippet(&expr.span);
+            }
+
+            // `..=` desugars into RangeInclusive::new(...)
+            ExprKind::Call(ref func, _) => {
+                if let ExprKind::Path(QPath::TypeRelative(ref ty, ref segment)) = func.node {
+                    if let TyKind::Path(QPath::Resolved(None, ref path)) = ty.node {
+                        let calls_new = segment.ident.as_str() == "new";
+
+                        return is_range_path(&path) && calls_new &&
+                            !is_range_struct_snippet(&expr.span);
+                    }
+                }
+            }
+
+            _ => {}
+        }
+
+        false
```

Comment:
Instead of explicitly returning `false` here, you could just move it inside the `match` (and get rid of the `return`s elsewhere).

Reply:
That would proliferate `else { false }` in `ExprKind::Call` match arm due to doubly-nested `if let`. Early `return` keeps it a bit shorter.

Reply:
Ah, that's true!